jerry-core/ecma/base/ecma-helpers-value.c

```diff
@@ -33,60 +30,90 @@ JERRY_STATIC_ASSERT (sizeof (ecma_value_t) * JERRY_BITSINBYTE >= ECMA_VALUE_SIZE
  */
 
 /**
+ * Masking the type and flags
+ */
+#define ECMA_VALUE_FULL_MASK (ECMA_VALUE_TYPE_MASK | ECMA_VALUE_ERROR_FLAG)
+
+JERRY_STATIC_ASSERT (ECMA_TYPE___MAX <= ECMA_VALUE_TYPE_MASK,
+                     ecma_types_must_be_less_than_mask);
+
+JERRY_STATIC_ASSERT ((ECMA_VALUE_FULL_MASK + 1) == (1 << ECMA_VALUE_SHIFT),
+                     ecma_value_part_must_start_after_flags);
+
+JERRY_STATIC_ASSERT (ECMA_VALUE_SHIFT <= MEM_ALIGNMENT_LOG,
+                     ecma_value_shift_must_be_less_than_or_equal_than_mem_alignment_log);
+
+/**
  * Get type field of ecma value
  *
  * @return type field
  */
-ecma_type_t __attr_pure___
+static inline ecma_type_t __attr_pure___ __attr_always_inline___
 ecma_get_value_type_field (ecma_value_t value) /**< ecma value */
 {
-  return (ecma_type_t) JRT_EXTRACT_BIT_FIELD (ecma_value_t, value,
-                                              ECMA_VALUE_TYPE_POS,
-                                              ECMA_VALUE_TYPE_WIDTH);
+  return value & ECMA_VALUE_TYPE_MASK;
 } /* ecma_get_value_type_field */
 
 /**
- * Get value field of ecma value
+ * Convert a pointer into an ecma value.
  *
- * @return value field
+ * @return ecma value
  */
-static uintptr_t __attr_pure___
-ecma_get_value_value_field (ecma_value_t value) /**< ecma value */
+static inline ecma_value_t __attr_pure___ __attr_always_inline___
+ecma_pointer_to_ecma_value (const void *ptr) /**< pointer */
 {
-  return (uintptr_t) JRT_EXTRACT_BIT_FIELD (ecma_value_t, value,
-                                            ECMA_VALUE_VALUE_POS,
-                                            ECMA_VALUE_VALUE_WIDTH);
-} /* ecma_get_value_value_field */
+  if (sizeof (void *) <= sizeof (ecma_value_t))
```

Comment:
There is a slight difference between `if (0)` and `#if 0`. You wrote the former, I prefer the latter.

`if (0)` *can* be optiized away (and usually is, I admit) but still has to be valid code (e.g., it must reference existing structure fields, which may not be there because of preproc guards used elsewhere -- not applicable to this case), while `#if 0` certainly *is* dead code that the complier does not even see. That's why the latter is used normally.

Even in the current code base, we (almost) never decide on type sizes with C code, but (almost) always with preprocessor conditionals. I've worked my way through 250+ occurrences of `sizeof` in jerry-core, and the only exception I've found is `ecma-helpers-external-pointers.c`, which uses `if (sizeof (x) < sizeof (y))` style. Which it shouldn't either, IMO.